### PR TITLE
Use TypeScript project references

### PR DESCRIPTION
**Pull request by felixfbecker**
We have many TS projects that partly reference each other.
Currently, TS has to double-compile the common files for every project.
This is noticeable when:
- Running `yarn all:typecheck`
- In CPU/RAM usage in VS Code, as we had 6 `tsc` compilers running concurrently
- In sluggish code intelligence in VS Code, as the compiler has to deal with a mega-project instead of relying on precompiled project outputs
- Node projects like `lsif` or `extrunner` cannot easily share code with `shared`

After this PR, we only run _one_ `tsc` in VS Code and `yarn all:typecheck`. For each project, compilation results are written to disk (`out` directories). If nothing changed in a project, it does not get rebuilt.
Code intelligence still jumps directly to the source as of TS 3.7.
Node projects like `lsif` or `extrunner` can easily share functions from `shared` with this by referencing the project (it's not referenced yet as there are no references yet).
It also enforces that we don't use circular references between projects.

This also changes our webpack configs and gulpfiles to be written in JS with `@ts-check` instead of TS. This saves multiple seconds of gulp tasks (which is run on every build and on `yarn install`) and makes it easier to use project references as those files don't have to be in a project. `@ts-check` provides a sufficient level of in-editor checking and editor experience, and these files are rarely changed, with mistakes usually resulting in build misconfiguration that wouldn't be caught by TS anyway.

### CODE REVIEW

File: browser/gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// @ts-check
```
Comment (eseliger) [inline browser/gulpfile.js]: would errors in this file be caught in CI?

File: dev/launch.sh [hunk 2]
```diff
@@ -118,6 +118,10 @@ export PATH="$PWD/.bin:$PWD/node_modules/.bin:$PATH"
     pushd ./lsif && yarn --no-progress && popd
 }
 
+# Build once to make sure editor codeintel works
+# This is fast if no changes were made.
+yarn run build-ts
```
Comment (eseliger) [inline dev/launch.sh]: Would it make sense to add the watcher for that task to the Procfile as we do for web, ...?
That way, after switching a branch you don't have to restart the dev env if you see errors and don't have the watcher running manually 🤔

Comment (felixfbecker) [reply dev/launch.sh]: That would mean that VS Code users still run the compiler twice (we have to run it in VS Code to get errors in the problems panel), which this PR tries to remove. Since this is only required for VS Code codeintel, not for the webpack build, I think it makes sense to have it only run automatically as a VS Code task. If someone needs it in their editor, similar configs could be added for those editors.

File: browser/gulpfile.js (hunk 1 above)
Comment (felixfbecker) [reply browser/gulpfile.js]: Depends on what errors. Errors that actually break the build, yes, because the build would fail or e2e tests would fail. Type errors where the build still succeeds, no. I think this is a good trade off

File: lsif/src/tests/linked-reference-results.test.ts [hunk 3]
```diff
@@ -18,7 +18,6 @@ describe('Backend', () => {
     it('should find all linked refs of `foo` in index.ts', async () => {
         if (!ctx.backend) {
             fail('failed beforeAll')
-            return
```
Comment (efritz) [inline lsif/src/tests/linked-reference-results.test.ts]: Why did you remove all of these? These stop spurious errors from printing and hiding a error in the setup.

Comment (felixfbecker) [reply lsif/src/tests/linked-reference-results.test.ts]: `fail()` always throws an exception and is therefor typed with return type `never`. In TS3.7, anything after a function call with return type `never` is (correctly) detected as unreachable code and therefor a compile error. This shouldn't change any behavior or hide errors from printing.

**Comment by unknwon** [comment]
Does this mean we will have shorter launch time for `./dev/launch.sh`? ☺️

File: schema/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "../tsconfig.json",
+  "references": [],
+  "compilerOptions": {
+    "rootDir": ".",
+    "outDir": "out",
```
Comment (unknwon) [inline schema/tsconfig.json]: I think we currently output to the same directory, so should this line be this?
```suggestion
    "outDir": ".",
```

Comment (felixfbecker) [reply schema/tsconfig.json]: No, currently we don't output at all. Using `.` would actually not work because the output here are `.d.ts` and the `.json` source files, so the output would override the input.

Comment (unknwon) [reply schema/tsconfig.json]: Ah OK, is this file required in this directory? It seems odd having a config file and we don't output anything.

Comment (felixfbecker) [reply schema/tsconfig.json]: Yes, because multiple other TypeScript projects import these files, and every imported TS file needs to belong to exactly one project. We actually do output something as of this PR (the .json and .d.ts files)

Comment (unknwon) [reply schema/tsconfig.json]: Cool, got it!

File: lsif/package.json [hunk 5]
```diff
@@ -73,15 +73,14 @@
     "source-map-support": "^0.5.16",
     "tsc-watch": "^4.0.0",
     "tslint": "^5.20.1",
-    "typescript": "^3.5.3",
```
Comment (efritz) [inline lsif/package.json]: I've done this in https://github.com/sourcegraph/sourcegraph/pull/6527 as well so I could get some of the newer TS features for a refactor. If this PR makes it in first I can close the other.

File: lsif/package.json [hunk 6]
```diff
@@ -73,15 +73,14 @@
...
+    "typescript": "^3.7.2",
     "typescript-json-schema": "^0.40.0"
   },
   "scripts": {
-    "build": "tsc",
+    "build": "tsc -b .",
     "postbuild": "yarn run copy-files",
     "copy-files": "copyfiles -u 1 src/**/*.lua out",
     "test": "jest",
-    "typecheck": "tsc --noEmit",
```
Comment (efritz) [inline lsif/package.json]: Why remove this option?

Comment (felixfbecker) [reply lsif/package.json]: Before, we ran this script in each project folder. Now we just build the root project which builds (and typechecks) everything. You can also run `build-ts` in this folder which does the same thing but emits (there is no benefit in having both).
